src/data/equiv/derangements/basic.lean

```diff
@@ -0,0 +1,265 @@
+begin
+  cases f_none : f none with a,
+  -- if f none is none, then the RHS is false, and this simplifies to "show f is not a derangement"
+  { simp only [is_derangement, foo, iff_false, not_forall, not_not],
+    use ⟨none, f_none⟩ },
+  -- otherwise we have to do actual proof
+  { simp only [foo, only_possible_fixed_point, is_derangement],
+    split,
+    -- (after contrapose): if (remove_none f) has a fixed point x ≠ a, then it's a fixed point
+    -- of f as well
+    { contrapose!,
+      rintro ⟨x, x_fixed, x_ne_a⟩,
+      rw remove_none_fixed_pt at x_fixed,
+      rcases x_fixed with (x_fixed | ⟨fnone_eq_x, fx_eq_none⟩),
+      -- either x was already a fixed point, which is easy...
+      { use ⟨some x, x_fixed⟩ },
+      -- ... or f swapped x and none, which is a contradiction
+      { exfalso,
+        have : some x = some a := by cc,
+        exact x_ne_a (option.some_inj.mp this) }},
+    -- (after contrapose): if f has a fixed point, then there is a fixed point of (remove_none f),
+    -- which isn't equal to a
+    { contrapose!,
+      rintro ⟨x, x_fixed⟩,
+      -- since f none = some a, we know x can't be none
+      cases x,
+      { exfalso,
+        exact option.some_ne_none a (by cc) },
+      -- so we have a fixed point (some x), it remains to show it's still a fixed point, and that
+      -- it's not equal to a
+      { use x,
+        rw remove_none_fixed_pt,
+        split,
+        { left, assumption },
+        { intro x_eq_a,
+          have : f (some x) = f none := by cc,
+          exact option.some_ne_none x (f.injective (this)) }}}}
+end
+
+-- TODO is the set better like this? or should i replace it with a union over (a : α)?
+def foo_set : set (option α × perm α) := {af : option α × perm α | foo af.fst af.snd}
+
+variables [decidable_eq α]
+
+lemma foo_set_eq : equiv.perm.decompose_option '' derangements (option α) = foo_set :=
+begin
+  -- rearrange it so that we have derangements (option α) = (...)⁻¹' foo_set
+  symmetry,
+  rw ← equiv.preimage_eq_iff_eq_image,
+  ext f,
+  exact (foo_lemma f).symm,
+end
+
+-- TODO this kinda makes more sense as a subset of `α × perm α`, but the proof was easier this way.
+/-- The set of derangements on `option α` is equivalent to the union of "permutations with either
+    no fixed points, `a` as the only fixed point", over all `a : α`. -/
+def derangements_equiv_sigma_opfp :
+  derangements (option α) ≃ Σ a : α, {f : perm α // only_possible_fixed_point f a} :=
+begin
+  -- TODO why do i need the @ to get it to infer α?
+  let fiber : option α → set (perm α) := λ a : option α, {f : perm α | (a, f) ∈ (@foo_set α)},
+
+  have fiber_imp_some : ∀ a : option α, (fiber a) → a.is_some,
+  { rintros a ⟨f, f_foo⟩,
+    cases a,
+    { exfalso, simp [foo_set, fiber] at f_foo, assumption },
+    { simp }},
+
+  have fiber_equiv_opfp : ∀ a : α, (fiber a) ≃ {f : perm α // only_possible_fixed_point f a},
+  { intro a, refl },
+
+  -- TODO how are calc blocks this big formatted?
```

Comment:
This is fine. Certainly align the equality symbols, as you've done. It's nice to align the `:`s, but it's okay to break this rule to save lines, as you've done.